File: 백준/Silver/24511. queuestack/queuestack.py
import sys
from collections import deque

input = sys.stdin.readline

N = int(input().strip())
A = list(map(int, input().strip().split()))
B = list(map(int, input().strip().split()))
M = int(input().strip())
C = list(map(int, input().strip().split()))

queues = deque()

for i in range(N):
    if A[i] == 0:
        queues.appendleft(B[i])

results = []

for value in C:
    queues.append(value)
    results.append(queues.popleft())

sys.stdout.write(' '.join(map(str, results)) + '\n')

File: 백준/Silver/24511. queuestack/queuestack.java
import java.io.*;
import java.util.*;
import java.util.stream.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        int N = Integer.parseInt(br.readLine());
        int[] A = Stream.of(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
        int[] B = Stream.of(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
        int M = Integer.parseInt(br.readLine());
        int[] C = Stream.of(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();

        Deque<Integer> queues = new ArrayDeque<>();

        for(int i = 0; i < N; i++){
            if(A[i] == 0) queues.offerFirst(B[i]);
        }

        StringBuilder sb = new StringBuilder();

        for(int value : C){
            queues.offerLast(value);
            sb.append(queues.pollFirst()).append(" ");
        }

        bw.write(sb.toString().trim());
        bw.newLine();
        bw.flush();
        bw.close();
        br.close();
    }
}